src/table/components/virtualized-table/hooks/use-get-hypercube-data-queue.ts

```diff
@@ -0,0 +1,67 @@
+import { useMemo, useRef } from 'react';
+
+export interface AbortablePromise {
+  isCancelled: boolean;
+}
+
+const pageToKey = ({ qLeft, qTop, qWidth, qHeight }: EngineAPI.INxPage) => `${qLeft}-${qTop}-${qWidth}-${qHeight}`;
+
+const useGetHyperCubeDataQueue = (
+  getDataPages: (qPages: EngineAPI.INxPage[]) => Promise<EngineAPI.INxDataPage[]>,
+  handleDataPages: (qDataPages: EngineAPI.INxDataPage[]) => void
+) => {
+  const queued = useRef(new Set<EngineAPI.INxPage>()); // Keep track of all unique pages that should be retrieved
+  const ongoing = useRef(new Set<EngineAPI.INxPage[]>()); // Keep track of ongoing request. Should be aborted if page info or layout is changed
+  const finished = useRef(new Set<string>()); // Keep track of all unqiue pages that have added to the queue
+
+  const queue = useMemo(
+    () => ({
+      enqueue: (page: EngineAPI.INxPage) => {
+        const key = pageToKey(page);
+
+        if (finished.current.has(key)) {
+          return;
+        }
+
+        queued.current.add(page);
+        // Add to finished before it's actually finished. That makes it easier to check if the page should be retrieved or not
+        // as otherwise all three sets (queued, ongoing and finished) have to be queried for presence.
+        finished.current.add(key);
+
+        if (queued.current.size === 1) {
```

Comment:
ok so trying to understand this. Whenever you run `enqueue,` you also add the passed `page` to `queued`. Then you check if there is only one entry in the queue. Can there be more than one? If so, isn't the rest of the queue skipped? or is it listening to something that makes it run later?

Reply:
It is assumed that the function that is calling `enqueue` is going to call it multiple times. As is the case in `useInfiniteScrollData`where `enqueue` is being called inside a for-loop.

`enqueue` is using the deterministic nature of the javascript event loop to "batch" queued pages.

```js
function task() {
  for (let index = 0; index < 10; index++) {
    // when the first index is passed to enqueue, `queueMicroTask` is called with a function, lets call that function "microTask" 
    enqueue(index);
  }
}

// In the event loop the execution flow would look something like this
task(); // Execute the task function
// and THEN when task() function has exited/finished the queue micro task is executed
microtask(); // Execute the micro task that now have 10 queued indexes as they where all added during the task() execution
```

Reply:
Yes there can and in most cases will be more then 1 queued page.